Add Elastic Fabric Adapter (EFA) metric collection to awscontainerinsightreceiver.

straussb opened the pull request:
Add Elastic Fabric Adapter (EFA) metric collection to awscontainerinsightreceiver.

The new component scrapes hardware counters from /sys/class/infiniband on disk.  The layout of that directory is:

```
/sys/class/infiniband/<device name>
└── ports
    └── 1
        └── hw_counters
            ├── rdma_read_bytes
            ├── rdma_write_bytes
            ├── rdma_write_recv_bytes
            ├── rx_bytes
            ├── rx_drops
            └── tx_bytes
```

These are cumulative counters and so they are converted to deltas before sending down the pipeline.
We sum up data from all ports.

The device data is joined with data from the kubelet podresources API which tells us which container a given device is
assigned to.

The metrics are reported at container, pod, and node levels.

This commit also refactors some metric decoration code from cadvisor to a common localnode decorator, intended to be
used by any awscontainerinsightreceiver component that gathers metrics from the local node (as oppoosed to e.g. the k8s
control plane scraper).  This is because we want to share the logic of populating PodName, Kubernetes labels, etc.

I also renamed RawContainerInsightsMetric to CIMetricImpl for brevity.

See also accompanying cloudwatch-agent PR: https://github.com/aws/amazon-cloudwatch-agent/pull/1098

movence commented on receiver/awscontainerinsightreceiver/internal/cadvisor/testutils/helpers.go:
```diff
@@ -64,13 +64,13 @@ func (m MockHostInfo) GetInstanceID() string {
 
 func (m MockHostInfo) GetInstanceType() string {
-	return "instance-id"
+	return "instance-type"
```
good catch 👍

movence commented on receiver/awscontainerinsightreceiver/internal/efa/efaSysfs.go:
```diff
@@ -0,0 +1,402 @@
+
+func calculateDelta(prev *metrics.MetricValue, val any, _ time.Time) (any, bool) {
+	var deltaValue uint64
+	if prev != nil {
```
nit: probably return the default(0) with false when it's null upfront. otherwise:
`return val.(uint64) - prev.RawValue.(uint64), true`

straussb replied:
Thanks, had copied this from somewhere else without thinking about it.  Your suggestion is much nicer.

movence commented on receiver/awscontainerinsightreceiver/internal/efa/efaSysfs.go:
```diff
@@ -0,0 +1,402 @@
+		s.fillMetric(nodeMetric, ci.TypeNodeEFA, containerInfo.Namespace, containerInfo.PodName,
+			containerInfo.ContainerName, string(deviceName), store.timestamp, counters)
+
+		for _, m := range []stores.CIMetric{containerMetric, podMetric, nodeMetric} {
```
you can merge this for loop into the loop where it calls `decorator.Decorate`. It's technically doing partial decoration on all metric types.

straussb replied:
I like to keep it separate so the series of steps is clear: 1) populate metric values, 2) add tags that the EFA scraper is responsible for, 3) pass to the decorator to fill in other metadata, and 4) convert to OTLP.

movence replied:
makes sense 👍

sam6134 commented on receiver/awscontainerinsightreceiver/internal/stores/localnode.go:
```diff
@@ -0,0 +1,184 @@
+			d.logger.Warn("Can't get containerInstanceId")
+		} else {
+			tags[ci.ContainerInstanceIDKey] = d.ecsInfo.GetContainerInstanceID()
+		}
```
Why cant we move lines 110-120 to addECSResources func as well

straussb replied:
Sure, done.  I had moved this from cadvisor_linux and didn't make any modifications.

sam6134 commented on receiver/awscontainerinsightreceiver/internal/stores/localnode.go:
```diff
@@ -0,0 +1,184 @@
+	}
+}
+
+func addECSResources(tags map[string]string) {
```
Might be a silly question, but is there any reason not to reuse this - https://github.com/amazon-contributing/opentelemetry-collector-contrib/blob/main/receiver/awscontainerinsightreceiver/internal/stores/utils.go#L98

straussb replied:
Done, moved the Instance* metric type logic to there.

straussb commented on receiver/awscontainerinsightreceiver/internal/efa/efaSysfs.go:
```diff
@@ -0,0 +1,402 @@
+
+func (s *Scraper) fillMetric(metric stores.CIMetric, metricType string, namespace string, podName string,
+	containerName string, deviceName string, timestamp time.Time, counters *efaCounters) {
+	s.assignRateValueToField(ci.MetricName(metricType, ci.EfaRdmaReadBytes), counters.rdmaReadBytes,
```
I think I did what you're asking in the next commit, let me know if you were imagining something different.

straussb commented on receiver/awscontainerinsightreceiver/internal/efa/efaSysfs.go:
```diff
@@ -0,0 +1,402 @@
+		case <-ticker.C:
+			err := s.scrape()
+			if err != nil {
+				// TODO only print this every so often
```
Removed and replaced with a check to skip everything if there is no EFA data on the host at all.

straussb commented:
For latest revision, I squashed all my commits together and rebased on aws-cwa-dev.

This got rid of the history... if you want to see the previous state of my PR, you can view the tree at: https://github.com/straussb/opentelemetry-collector-contrib/tree/db5503cb45f8e78b169da51c95bc4489f55db3ff

And compare to current state at: https://github.com/straussb/opentelemetry-collector-contrib/compare/db5503cb45f8e78b169da51c95bc4489f55db3ff..efa

The files that had rebase conflicts were:
* internal/aws/containerinsight/const.go
* receiver/awscontainerinsightreceiver/internal/prometheusscraper/decoratorconsumer/decorator.go
* receiver/awscontainerinsightreceiver/internal/stores/utils.go
* receiver/awscontainerinsightreceiver/internal/stores/utils_test.go
* receiver/awscontainerinsightreceiver/receiver.go

So review those a bit more carefully.

Since last review, I also added a bit more logic in [efaSysfs.go](https://github.com/amazon-contributing/opentelemetry-collector-contrib/pull/186/files#diff-f64355eaa5a122b4dc6f283dd141bae4494b6d4ed00c5a4fd41d7e519f31e2e6R337) and [podresourcesclient.go](https://github.com/amazon-contributing/opentelemetry-collector-contrib/pull/186/files#diff-3cfdd14b953421922da05b8bc8812e63cc88f817630ed0e3deecf7ffe8f1946bR69) to check permissions/ownership of `/sys/class/infiniband` and the kubelet podresources socket, for security.

movence commented on receiver/awscontainerinsightreceiver/internal/efa/efaSysfs.go:
```diff
@@ -0,0 +1,443 @@
+	// This was already patched and submitted, see
+	// https://www.spinics.net/lists/linux-rdma/msg68596.html
+	// Remove this as soon as the fix lands in the enterprise distros.
+	if strings.Contains(stringValue, "N/A (no PMA)") {
```
will this break if the fix gets promoted? how do we track this change promotion?

straussb replied:
Good question.  I'm going to follow up with a commit to improve partial error handling - still post node-level metrics if we couldn't get pod resources data, as we discussed in the meeting; and also include data for some devices/counters even if reading some other data fails.  As part of that commit, I will remove this logic and just skip if parsing to int failed.

straussb commented on internal/aws/containerinsight/const.go:
```diff
@@ -158,11 +166,14 @@ const (
+	TypePodGPU          = "PodGPU"
+	TypeNodeGPU         = "NodeGPU"
+	TypeClusterGPU      = "ClusterGPU"
 	TypeNeuronContainer = "ContainerNeuron"
```
Thanks, done.

straussb commented on receiver/awscontainerinsightreceiver/internal/efa/efaSysfs.go:
```diff
@@ -0,0 +1,443 @@
+
+	stat, ok := info.Sys().(*syscall.Stat_t)
+	if !ok {
+		return true, nil
```
This would essentially mean that we are not running on Linux.  But you're right, we should skip in that case so we remember to implement this logic for Windows if we ever add support for that.  Changed in next commit.

straussb commented on receiver/awscontainerinsightreceiver/receiver.go:
```diff
@@ -111,26 +121,33 @@ func (acir *awsContainerInsightReceiver) Start(ctx context.Context, host compone
 				acir.settings.Logger.Debug("Unable to start neuron scraper", zap.Error(err))
 			}
+
+			acir.initEfaSysfsScraper(localnodeDecorator)
```
Previously there was no reason that the scraper couldn't start.  Added one case now.

straussb commented on receiver/awscontainerinsightreceiver/internal/efa/efaSysfs.go:
```diff
@@ -0,0 +1,443 @@
+	ReadCounter(deviceName efaDeviceName, port string, counter string) (uint64, error)
+}
+
+type podResourcesStore interface {
```
I think it's OK to duplicate in this case - I think the typical Go style would say that each class can independently mock out its dependencies as it chooses to.
